backend/danswer/search/pipeline.py

```diff
@@ -48,15 +77,148 @@ def __init__(
+                                # can be assumed to be safe
+                                IndexFilters(access_control_list=None),
+                            ),
+                        )
+                    )
+
+            list_inference_chunks = run_functions_tuples_in_parallel(
+                functions_with_args, allow_failures=False
+            )
+
+            for chunk in unique_chunks:
+                inf_chunks = list_inference_chunks.pop(0)
+                combined_content = "\n".join([chunk.content for chunk in inf_chunks])
+                final_inference_sections.append(
+                    InferenceSection.from_chunk(chunk, content=combined_content)
+                )
+
+            return final_inference_sections
+
+        # General flow:
+        # - Combine chunks into lists by document_id
+        # - For each document, run merge-intervals to get combined ranges
+        # - Fetch all of the new chunks with contents for the combined ranges
+        # - Map it back to the combined ranges (which each know their "center" chunk)
+        # - Reiterate the chunks again and map to the results above based on the chunk.
+        #   This maintains the original chunks ordering. Note, we cannot simply sort by score here
+        #   as reranking flow may wipe the scores for a lot of the chunks.
+        doc_chunk_ranges_map = defaultdict(list)
+        for chunk in chunks:
+            doc_chunk_ranges_map[chunk.document_id].append(
+                ChunkRange(
+                    chunk=chunk,
+                    start=max(0, chunk.chunk_id - self._search_query.chunks_above),
+                    # No max known ahead of time, filter will handle this anyway
+                    end=chunk.chunk_id + self._search_query.chunks_below,
+                )
+            )
+
+        merged_ranges = [
+            merge_chunk_intervals(ranges) for ranges in doc_chunk_ranges_map.values()
+        ]
+        reverse_map = {r.chunk: r for doc_ranges in merged_ranges for r in doc_ranges}
+
+        for chunk_range in reverse_map.values():
+            functions_with_args.append(
+                (
+                    self.document_index.id_based_retrieval,
+                    (
+                        chunk_range.chunk.document_id,
+                        chunk_range.start,
+                        chunk_range.end,
+                        # There is no chunk level permissioning, this expansion around chunks
+                        # can be assumed to be safe
+                        IndexFilters(access_control_list=None),
+                    ),
+                )
+            )
+
+        # list of list of inference chunks where the inner list needs to be combined for content
+        list_inference_chunks = run_functions_tuples_in_parallel(
+            functions_with_args, allow_failures=False
+        )
+
+        for chunk_range in reverse_map.values():
+            inf_chunks = list_inference_chunks.pop(0)
```

Comment:
Can we pop from the back? I believe popping from the front is a bit slower: https://stackoverflow.com/a/34633242/22737043